File: SDK/CentralService/python/examples/e2e.py
from __future__ import annotations

import os
import sys
import time
import uuid
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
for project_root in (ROOT / "service", ROOT / "client"):
    project_root_str = str(project_root)
    if project_root_str not in sys.path:
        sys.path.insert(0, project_root_str)

from erp_centralservice_client import CentralServiceDiscoveryClient, calculate_network_score, load_sdk_options_from_env
from erp_centralservice_service import CentralServiceServiceClient


def get_timeout_ms() -> int:
    raw_value = os.environ.get("CENTRAL_SERVICE_TIMEOUT_MS") or os.environ.get("CENTRAL_SERVICE_E2E_TIMEOUT_MS") or "5000"
    try:
        value = int(raw_value)
    except ValueError:
        value = 5000
    return value if value > 0 else 5000


def get_timeout_sec() -> float:
    return get_timeout_ms() / 1000.0


def get_break_wait_seconds() -> float:
    raw_value = os.environ.get("CENTRAL_SERVICE_E2E_BREAK_WAIT_SECONDS") or "65"
    try:
        value = float(raw_value)
    except ValueError:
        value = 65.0
    return value if value > 0 else 65.0


def get_scenario() -> str:
    return os.environ.get("CENTRAL_SERVICE_E2E_SCENARIO", "smoke")


def get_service_name() -> str:
    return os.environ.get("CENTRAL_SERVICE_E2E_SERVICE_NAME", "SdkE2E")


def get_service_port() -> int:
    raw_value = os.environ.get("CENTRAL_SERVICE_E2E_SERVICE_PORT") or "18083"
    try:
        value = int(raw_value)
    except ValueError:
        value = 18083
    return value if value > 0 else 18083


def get_endpoints() -> list[dict]:
    options = load_sdk_options_from_env(timeout_sec=get_timeout_sec())
    endpoints = options.get("endpoints")
    if isinstance(endpoints, list) and endpoints:
        return endpoints
    return [{"baseUrl": options.get("base_url")}]


def create_single_endpoint_options(endpoint: dict) -> dict:
    return {"base_url": endpoint["baseUrl"], "timeout_sec": get_timeout_sec()}


def create_registration(service_id: str) -> dict:
    return {
        "id": service_id,
        "name": get_service_name(),
        "host": "127.0.0.1",
        "localIp": "127.0.0.1",
        "operatorIp": "127.0.0.1",
        "publicIp": "127.0.0.1",
        "port": get_service_port(),
        "serviceType": "Web",
        "healthCheckUrl": "/health",
        "healthCheckPort": 0,
        "heartbeatIntervalSeconds": 0,
        "weight": 1,
        "metadata": {"sdk": "python", "scenario": get_scenario()},
    }


def create_stable_service_id() -> str:
    return os.environ.get("CENTRAL_SERVICE_E2E_SERVICE_ID") or str(uuid.uuid4())


def assert_condition(condition: bool, message: str) -> None:
    if not condition:
        raise RuntimeError(f"[{get_scenario()}] {message}")


def assert_optional_expected_id(step_name: str, actual_id: str) -> None:
    env_name = f"CENTRAL_SERVICE_E2E_EXPECTED_{step_name.upper()}_ID"
    expected_id = os.environ.get(env_name)
    if expected_id:
        assert_condition(actual_id == expected_id, f"{step_name} 期望 id={expected_id}，实际={actual_id}")


def deregister_quietly(client: CentralServiceServiceClient, service_id: str) -> None:
    if not service_id:
        return
    try:
        client.deregister(service_id)
    except Exception:
        pass


def run_smoke() -> None:
    endpoint = get_endpoints()[0]
    service = CentralServiceServiceClient(create_single_endpoint_options(endpoint))
    discovery = CentralServiceDiscoveryClient(create_single_endpoint_options(endpoint))
    service_id = ""
    try:
        reg = service.register(create_registration(""))
        service_id = reg["id"]
        print(f"[py][smoke] registered id={service_id}")
        listed = discovery.list(get_service_name())
        assert_condition(any(item.get("id") == service_id for item in listed.get("services") or []), "list 未包含刚注册的服务")
        best = discovery.discover_best(get_service_name())
        assert_condition(bool(best.get("id")), "discover_best 未返回服务实例")
        evaluated = discovery.evaluate_network(service_id)
        print(f"[py][smoke] network evaluated score={calculate_network_score(evaluated)}")
        all_net = discovery.get_network_all()
        assert_condition(isinstance(all_net, list), "network/all 未返回数组")
    finally:
        deregister_quietly(service, service_id)


def run_service_fanout() -> None:
    endpoints = get_endpoints()
    assert_condition(len(endpoints) >= 2, "service_fanout 至少需要 2 个端点")
    service_id = create_stable_service_id()
    sessions: list[tuple[dict, CentralServiceServiceClient, str]] = []
    try:
        for endpoint in endpoints:
            client = CentralServiceServiceClient(create_single_endpoint_options(endpoint))
            reg = client.register(create_registration(service_id))
            sessions.append((endpoint, client, reg["id"]))

        for endpoint, _, registered_id in sessions:
            assert_condition(registered_id == service_id, f"端点 {endpoint['baseUrl']} 未复用同一个 serviceId")
            discovery = CentralServiceDiscoveryClient(create_single_endpoint_options(endpoint))
            listed = discovery.list(get_service_name())
            assert_condition(any(item.get("id") == service_id for item in listed.get("services") or []), f"端点 {endpoint['baseUrl']} 未查询到注册结果")
            best = discovery.discover_best(get_service_name())
            assert_condition(best.get("id") == service_id, f"端点 {endpoint['baseUrl']} discover_best 返回了错误实例")
    finally:
        for _, client, registered_id in sessions:
            deregister_quietly(client, registered_id)


def run_transport_failover() -> None:
    discovery = CentralServiceDiscoveryClient(load_sdk_options_from_env(timeout_sec=get_timeout_sec()))
    result = discovery.discover_best(get_service_name())
    assert_condition(bool(result.get("id")), "transport_failover 未返回服务实例")
    assert_optional_expected_id("first", result.get("id", ""))


def run_business_no_failover() -> None:
    discovery = CentralServiceDiscoveryClient(load_sdk_options_from_env(timeout_sec=get_timeout_sec()))
    endpoints = get_endpoints()
    try:
        discovery.discover_best(get_service_name())
        raise RuntimeError(f"[{get_scenario()}] 期望业务失败，但调用成功了")
    except Exception as exc:
        kind = getattr(exc, "kind", "")
        assert_condition(kind != "Transport", "业务失败场景不应被识别为传输失败")
        assert_condition(endpoints[0]["baseUrl"] in str(exc), "错误信息未包含主端点上下文")
        print(f"[py][business_no_failover] kind={kind}")


def run_max_attempts() -> None:
    discovery = CentralServiceDiscoveryClient(load_sdk_options_from_env(timeout_sec=get_timeout_sec()))
    result = discovery.discover_best(get_service_name())
    assert_condition(bool(result.get("id")), "max_attempts 未返回服务实例")
    assert_optional_expected_id("first", result.get("id", ""))


def run_circuit_open() -> None:
    discovery = CentralServiceDiscoveryClient(load_sdk_options_from_env(timeout_sec=get_timeout_sec()))
    first = discovery.discover_best(get_service_name())
    second = discovery.discover_best(get_service_name())
    assert_condition(bool(first.get("id")) and bool(second.get("id")), "circuit_open 未返回有效实例")
    assert_optional_expected_id("first", first.get("id", ""))
    assert_optional_expected_id("second", second.get("id", ""))


def run_circuit_recovery() -> None:
    discovery = CentralServiceDiscoveryClient(load_sdk_options_from_env(timeout_sec=get_timeout_sec()))
    first = discovery.discover_best(get_service_name())
    time.sleep(get_break_wait_seconds())
    second = discovery.discover_best(get_service_name())
    third = discovery.discover_best(get_service_name())
    assert_condition(
        bool(first.get("id")) and bool(second.get("id")) and bool(third.get("id")),
        "circuit_recovery 未返回有效实例",
    )
    assert_optional_expected_id("first", first.get("id", ""))
    assert_optional_expected_id("second", second.get("id", ""))
    assert_optional_expected_id("third", third.get("id", ""))


def run_half_open_reopen() -> None:
    discovery = CentralServiceDiscoveryClient(load_sdk_options_from_env(timeout_sec=get_timeout_sec()))
    first = discovery.discover_best(get_service_name())
    time.sleep(get_break_wait_seconds())
    second = discovery.discover_best(get_service_name())
    third = discovery.discover_best(get_service_name())
    assert_condition(bool(first.get("id")) and bool(second.get("id")) and bool(third.get("id")), "half_open_reopen 未返回有效实例")
    assert_optional_expected_id("first", first.get("id", ""))
    assert_optional_expected_id("second", second.get("id", ""))
    assert_optional_expected_id("third", third.get("id", ""))


def main() -> None:
    endpoints = get_endpoints()
    print(f"[py] scenario={get_scenario()}")
    print(f"[py] timeoutMs={get_timeout_ms()}")
    print(f"[py] endpoints={endpoints}")

    handlers = {
        "smoke": run_smoke,
        "service_fanout": run_service_fanout,
        "transport_failover": run_transport_failover,
        "business_no_failover": run_business_no_failover,
        "max_attempts": run_max_attempts,
        "circuit_open": run_circuit_open,
        "circuit_recovery": run_circuit_recovery,
        "half_open_reopen": run_half_open_reopen,
    }

    handler = handlers.get(get_scenario())
    assert_condition(handler is not None, f"不支持的场景: {get_scenario()}")
    handler()
    print(f"[py] scenario passed: {get_scenario()}")


if __name__ == "__main__":
    try:
        main()
    except Exception as exc:
        print(f"[py] scenario failed: {get_scenario()} {exc}", file=sys.stderr)
        raise

File: SDK/CentralService/javascript/examples/e2e.js
const crypto = require("crypto");
const { CentralServiceServiceClient, loadCentralServiceOptionsFromEnv } = require("../service");
const {
  CentralServiceDiscoveryClient,
  calculateNetworkScore,
} = require("../client");

function getScenario() {
  return process.env.CENTRAL_SERVICE_E2E_SCENARIO || "smoke";
}

function getTimeoutMs() {
  const value = Number(process.env.CENTRAL_SERVICE_TIMEOUT_MS || process.env.CENTRAL_SERVICE_E2E_TIMEOUT_MS || 5000);
  return Number.isFinite(value) && value > 0 ? value : 5000;
}

function getBreakWaitMs() {
  const value = Number(process.env.CENTRAL_SERVICE_E2E_BREAK_WAIT_SECONDS || 65);
  return (Number.isFinite(value) && value > 0 ? value : 65) * 1000;
}

function getServiceName() {
  return process.env.CENTRAL_SERVICE_E2E_SERVICE_NAME || "SdkE2E";
}

function getPort() {
  const value = Number(process.env.CENTRAL_SERVICE_E2E_SERVICE_PORT || 18082);
  return Number.isFinite(value) && value > 0 ? value : 18082;
}

function loadOptions() {
  return loadCentralServiceOptionsFromEnv({ timeoutMs: getTimeoutMs() });
}

function getEndpoints() {
  const options = loadOptions();
  return Array.isArray(options.endpoints) && options.endpoints.length > 0
    ? options.endpoints
    : [{ baseUrl: options.baseUrl }];
}

function createSingleEndpointOptions(endpoint) {
  return {
    baseUrl: endpoint.baseUrl,
    timeoutMs: getTimeoutMs(),
  };
}

function createRegistration(serviceId) {
  return {
    id: serviceId,
    name: getServiceName(),
    host: "127.0.0.1",
    localIp: "127.0.0.1",
    operatorIp: "127.0.0.1",
    publicIp: "127.0.0.1",
    port: getPort(),
    serviceType: "Web",
    healthCheckUrl: "/health",
    healthCheckPort: 0,
    heartbeatIntervalSeconds: 0,
    weight: 1,
    metadata: { sdk: "javascript", scenario: getScenario() },
  };
}

function createStableServiceId() {
  return process.env.CENTRAL_SERVICE_E2E_SERVICE_ID || crypto.randomUUID();
}

function assertCondition(condition, message) {
  if (!condition) {
    throw new Error("[" + getScenario() + "] " + message);
  }
}

function assertOptionalExpectedId(stepName, actualId) {
  const envName = "CENTRAL_SERVICE_E2E_EXPECTED_" + stepName.toUpperCase() + "_ID";
  const expectedId = process.env[envName];
  if (expectedId) {
    assertCondition(actualId === expectedId, stepName + " 期望 id=" + expectedId + "，实际=" + actualId);
  }
}

async function registerOnEndpoint(endpoint, serviceId) {
  const client = new CentralServiceServiceClient(createSingleEndpointOptions(endpoint));
  const registration = await client.register(createRegistration(serviceId));
  return { client, serviceId: registration.id };
}

async function deregisterQuietly(client, serviceId) {
  if (!client || !serviceId) return;
  try {
    await client.deregister(serviceId);
  } catch {
  }
}

async function runSmoke() {
  const endpoint = getEndpoints()[0];
  const service = new CentralServiceServiceClient(createSingleEndpointOptions(endpoint));
  const discovery = new CentralServiceDiscoveryClient(createSingleEndpointOptions(endpoint));
  let serviceId = "";
  try {
    const reg = await service.register(createRegistration(""));
    serviceId = reg.id;
    console.log("[js][smoke] registered id=" + serviceId);
    const listed = await discovery.list(getServiceName());
    assertCondition((listed?.services || []).some((item) => item.id === serviceId), "list 未包含刚注册的服务");
    const best = await discovery.discoverBest(getServiceName());
    assertCondition(Boolean(best?.id), "discoverBest 未返回服务 id");
    const evaluated = await discovery.evaluateNetwork(serviceId);
    console.log("[js][smoke] network evaluated score=" + calculateNetworkScore(evaluated));
    const all = await discovery.getNetworkAll();
    assertCondition(Array.isArray(all), "network/all 未返回数组");
  } finally {
    await deregisterQuietly(service, serviceId);
  }
}

async function runServiceFanout() {
  const endpoints = getEndpoints();
  assertCondition(endpoints.length >= 2, "service_fanout 至少需要 2 个端点");
  const serviceId = createStableServiceId();
  const sessions = [];
  try {
    for (const endpoint of endpoints) {
      const session = await registerOnEndpoint(endpoint, serviceId);
      sessions.push({ ...session, endpoint });
    }

    for (const session of sessions) {
      assertCondition(session.serviceId === serviceId, "端点 " + session.endpoint.baseUrl + " 未复用同一个 serviceId");
      const discovery = new CentralServiceDiscoveryClient(createSingleEndpointOptions(session.endpoint));
      const listed = await discovery.list(getServiceName());
      assertCondition((listed?.services || []).some((item) => item.id === serviceId), "端点 " + session.endpoint.baseUrl + " 未查询到注册结果");
      const best = await discovery.discoverBest(getServiceName());
      assertCondition(best?.id === serviceId, "端点 " + session.endpoint.baseUrl + " discoverBest 返回了错误实例");
    }
  } finally {
    for (const session of sessions) {
      await deregisterQuietly(session.client, session.serviceId);
    }
  }
}

async function runTransportFailover() {
  const discovery = new CentralServiceDiscoveryClient(loadOptions());
  const result = await discovery.discoverBest(getServiceName());
  assertCondition(Boolean(result?.id), "transport_failover 未返回服务实例");
  assertOptionalExpectedId("first", result.id);
}

async function runBusinessNoFailover() {
  const discovery = new CentralServiceDiscoveryClient(loadOptions());
  const endpoints = getEndpoints();
  try {
    await discovery.discoverBest(getServiceName());
    throw new Error("[" + getScenario() + "] 期望业务失败，但调用成功了");
  } catch (error) {
    assertCondition(error && error.kind !== "Transport", "业务失败场景不应被识别为传输失败");
    assertCondition(String(error.message || "").includes(endpoints[0].baseUrl), "错误信息未包含主端点上下文");
    console.log("[js][business_no_failover] kind=" + error.kind);
  }
}

async function runMaxAttempts() {
  const discovery = new CentralServiceDiscoveryClient(loadOptions());
  const result = await discovery.discoverBest(getServiceName());
  assertCondition(Boolean(result?.id), "max_attempts 未返回服务实例");
  assertOptionalExpectedId("first", result.id);
}

async function runCircuitOpen() {
  const discovery = new CentralServiceDiscoveryClient(loadOptions());
  const first = await discovery.discoverBest(getServiceName());
  const second = await discovery.discoverBest(getServiceName());
  assertCondition(Boolean(first?.id) && Boolean(second?.id), "circuit_open 未返回有效实例");
  assertOptionalExpectedId("first", first.id);
  assertOptionalExpectedId("second", second.id);
}

async function runCircuitRecovery() {
  const discovery = new CentralServiceDiscoveryClient(loadOptions());
  const first = await discovery.discoverBest(getServiceName());
  await new Promise((resolve) => setTimeout(resolve, getBreakWaitMs()));
  const second = await discovery.discoverBest(getServiceName());
  const third = await discovery.discoverBest(getServiceName());
  assertCondition(Boolean(first?.id) && Boolean(second?.id) && Boolean(third?.id), "circuit_recovery 未返回有效实例");
  assertOptionalExpectedId("first", first.id);
  assertOptionalExpectedId("second", second.id);
  assertOptionalExpectedId("third", third.id);
}

async function runHalfOpenReopen() {
  const discovery = new CentralServiceDiscoveryClient(loadOptions());
  const first = await discovery.discoverBest(getServiceName());
  await new Promise((resolve) => setTimeout(resolve, getBreakWaitMs()));
  const second = await discovery.discoverBest(getServiceName());
  const third = await discovery.discoverBest(getServiceName());
  assertCondition(Boolean(first?.id) && Boolean(second?.id) && Boolean(third?.id), "half_open_reopen 未返回有效实例");
  assertOptionalExpectedId("first", first.id);
  assertOptionalExpectedId("second", second.id);
  assertOptionalExpectedId("third", third.id);
}

async function main() {
  const options = loadOptions();
  const endpoints = getEndpoints();
  console.log("[js] scenario=" + getScenario());
  console.log("[js] baseUrl=" + options.baseUrl);
  console.log("[js] endpoints=" + JSON.stringify(endpoints));

  const handlers = {
    smoke: runSmoke,
    service_fanout: runServiceFanout,
    transport_failover: runTransportFailover,
    business_no_failover: runBusinessNoFailover,
    max_attempts: runMaxAttempts,
    circuit_open: runCircuitOpen,
    circuit_recovery: runCircuitRecovery,
    half_open_reopen: runHalfOpenReopen,
  };

  const handler = handlers[getScenario()];
  assertCondition(Boolean(handler), "不支持的场景: " + getScenario());
  await handler();
  console.log("[js] scenario passed: " + getScenario());
}

main().catch((err) => {
  console.error("[js] scenario failed:", getScenario(), err && err.stack ? err.stack : err);
  process.exitCode = 1;
});
